datafusion/core/src/physical_optimizer/sort_pushdown.rs

```diff
@@ -176,6 +176,7 @@ fn pushdown_requirement_to_children(
         || plan.as_any().is::<ProjectionExec>()
         || is_limit(plan)
         || plan.as_any().is::<HashJoinExec>()
+        || pushdown_would_violate_requirements(parent_required, plan.as_ref())
```

Comment:
In general it would be great to avoid these explicit checks for types over time in favor of functions on the ExecutionPlan nodes